File: src/StompFrame.cpp
#include "StompFrame.h"

StompFrame::StompFrame() {}
StompFrame::StompFrame(string input) {
    int loc = input.find('\n');
    command = input.substr(0, loc);
    input = input.substr(loc + 1, input.length() - loc + 2);
    while(input.at(0) != '\n'){
        loc = input.find('\n');
        int loc2 = input.find(':');
        string tmp = input.substr(loc2, string::npos);
        headers.insert(make_pair(input.substr(0, loc2), tmp.substr(1 , tmp.find('\n') - 1)));
        input = input.substr(loc + 1);
    }
    input = input.substr(1);
    if(input.length() != 0 && input.at(input.length()-1) == '\n')
        input = input.substr(0, input.length() - 1);
    body = input;
}
StompFrame:: StompFrame(string _command, unordered_map<string, string> _headers, string _body): command(_command), headers(_headers), body(_body) {
}

StompFrame::~StompFrame() {

}

string StompFrame::getCommand() {
    return command;
}

unordered_map<string,string >& StompFrame:: getHeaders() {
    return headers;
}
string StompFrame:: getBody(){
    return body;
}
void StompFrame:: setBody(string _body) {
    body = _body;
}
void StompFrame:: setCommand(string _command){
    command = _command;
}
void StompFrame:: setHeaders(unordered_map<string, string> _headers){
    headers = _headers;
}
void StompFrame:: addHeader(pair<string, string> header){
    headers.insert(header);
}
string StompFrame:: getHeader(string key){
    return headers[key];
}

string StompFrame:: getHeaderMap(){
    string output = "";
    for(auto& key : headers){
        output += key.first + ":" + key.second + '\n';
    }
    output += '\n';
    return output;
}

string StompFrame:: toString(){
    string output = getCommand() + '\n';
    output += getHeaderMap() + '\n' + '\n';
    output += getBody() + '\n';
    return output;
}

File: Server/src/main/java/bgu/spl/net/impl/stomp/StompFrame.java
package bgu.spl.net.impl.stomp;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class StompFrame {
    private String command;
    private Map<String, String> headerMap = new HashMap<>();
    private String body;

    public StompFrame() {
    }

    public StompFrame(String input) {
        int i = input.indexOf('\n');
        command = input.substring(0, i);
        input = input.substring(i + 1);
        while (input.charAt(0) != '\n') {
            i = input.indexOf('\n');
            int j = input.indexOf(':');
            headerMap.put(input.substring(0, j), input.substring(j + 1, i));
            input = input.substring(i + 1);
        }
        input = input.substring(1);
        if(input.length() != 0 && input.charAt(input.length()-1) == '\n')
            input = input.substring(0, input.length() - 1);
        body = input;
    }

    public StompFrame(String _command, HashMap<String, String> _headerMap, String _body){
        command = _command;
        headerMap = _headerMap;
        body = _body;
    }

    public String toString() {
        String output = getCommand() + '\n' + getHeaderMap() + getBody() + '\0';
        return output;
    }

    public String getCommand(){
        return command;
    }

    public String getHeaderMap() {
        String output = "";
        Set<String> keySet = headerMap.keySet();
        for (String key : keySet) {
            output = output + key + ":" + headerMap.get(key) + '\n';
        }
        output = output + '\n';
        return output;
    }

    public String getBody(){
        String output = body + '\n';
        return output;
    }

    public String getHeader(String key){
        return headerMap.get(key);
    }

    public void setCommand(String command) {
        this.command = command;
    }

    public void setHeaderMap(Map<String, String> headerMap) {
        this.headerMap = headerMap;
    }

    public void setHeader(String key, String value) {
        headerMap.put(key, value);
    }

    public void setBody(String body) {
        this.body = body;
    }

    /*public StompFrame connectedFrame(String header){
        StompFrame frame = new StompFrame();
        frame.setCommand("CONNECTED");
        HashMap<String, String> map = new HashMap<>();
        map.put("version", header);
        setHeaderMap(map);
        setBody("");
        return frame;
    }

    public StompFrame errorFrame(String error, StompFrame message, String explanation){
        StompFrame frame = new StompFrame();
        frame.setCommand("ERROR");
        HashMap<String,String> map = new HashMap<>();
        map.put("message", "error");
        if(message.getHeader("receipt") != null){
            map.put("receipt-id",message.getHeader("receipt"));
        }
        frame.setHeaderMap(map);
        frame.setBody("The message:" + '\n' + "-----" + message.toString().substring(0,message.toString().length()-1) +
                "-----" + '\n' + explanation + '\n' + '\0');
        return frame;
    }

    public StompFrame receiptFrame(StompFrame message){
        StompFrame frame = new StompFrame();
        frame.setCommand("RECEIPT");
        HashMap<String, String> map = new HashMap<>();
        map.put("receipt-id", message.getHeader("receipt"));
        frame.setHeaderMap(map);
        frame.setBody("");
        return frame;
    }

    public StompFrame messageFrame(String _body, Integer subscription, Integer id, Integer destination){
        StompFrame frame = new StompFrame();
        frame.setCommand("MESSAGE");
        HashMap<String, String> map = new HashMap<>();
        map.put("subscription", subscription.toString());
        map.put("Message-id", id.toString());
        map.put("destination", destination.toString());
        frame.setHeaderMap(map);
        frame.setBody(_body);
        return frame;
    }*/
}
